### PR TITLE
Added RecursiveBinarySearch [ C ]

### LINKED ISSUE #247: Recursive binary search [ C ]
There is no implementation of recursive binary search in c/c++.
Description - [https://en.wikipedia.org/wiki/Binary_search_algorithm](url)

### PR DESCRIPTION
**Pull request by HarshKhatore**
A program in C++ to perform a recursive binary search.

Fixes #247 

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous timespan of 7 days.

### CODE REVIEW

**Comment by singhpratyush** [comment]
Hi @HarshKhatore. Thanks for your PR.

We're glad that you showed interest in contributing to this PR.

Please stick with reviewers and coding standards to get this PR merged. This may take some time. Also, add an entry in `README.md` about this algorithm.

**Comment by HarshKhatore** [comment]
Hello @singhpratyush, sure I will add about the algorithm in the README.md 
Can you guide me on doing that?

**Comment by singhpratyush** [comment]
Just look at the "algorithms" table and add an entry (similar to other that are already there) to the appropriate column. Make sure that the link that you mention is correct.

Also, since we already have an iterative variant of binary search in that directory, please add another directory with a recursive suffix in it.

**Comment by HarshKhatore** [comment]
@singhpratyush , updated README.md and directory as required.

File: recursive_binary_search/binary_search.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
+/* Program - Binary search over an array
+   Author - Harsh Khatore
```
Comment (Monal5031) [inline recursive_binary_search/binary_search.cpp]: Please remove these top 5 lines.

File: recursive_binary_search/binary_search.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+using namespace std;
+
+/* function to perform binary search
+input parameters:
+ar - user-input array.
+l - left index of array.
+r - right index of array.
+ele - element to be searched.
+*/
+int binarySearch(int *ar, int l, int r, int ele)
```
Comment (Monal5031) [inline recursive_binary_search/binary_search.cpp]: Please correct indentations, refer other codes for it.

File: recursive_binary_search/binary_search.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+{
+   if (r >= l)
+   {
+        int mid = l + (r - l)/2;    //calculate mid point of the array
+ 
+        // If the element is present at the middle itself
+        // +1 so that it is easier to read
+        if (ar[mid] == ele)  return mid+1;
```
Comment (Monal5031) [inline recursive_binary_search/binary_search.cpp]: Add a block here, even though it has 1 statement. Do the same in other parts of code.

File: recursive_binary_search/binary_search.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+}
+ 
+int main(void)
+{
+   int n;   //number of elements in array.
+   cin >> n;
+   
+   int *ar = new int[n];    //dynamically allocating array for user-input
+   
+   for(int i=0;i<n;i++)
```
Comment (Monal5031) [inline recursive_binary_search/binary_search.cpp]: Hardcode inputs, don't take them from user.

File: recursive_binary_search/binary_search.cpp (hunk 1 above)
Comment (HarshKhatore) [inline recursive_binary_search/binary_search.cpp]: @Monal5031 OK

File: recursive_binary_search/binary_search.cpp (hunk 4 above)
Comment (HarshKhatore) [inline recursive_binary_search/binary_search.cpp]: @Monal5031 Ok

**Comment by HarshKhatore** [comment]
@singhpratyush, @Monal5031 done the required changes, kindly review.

File: binary_search/binary_search.c [hunk 5]
```diff
@@ -5,6 +5,37 @@
...
  */
+ 
+ /* function to perform recursive binary search
+ input parameters:
+ ar - user-input array.
+ l - left index of array.
+ r - right index of array.
+ ele - element to be searched.
+ */
+ int binarySearch(const int *ar, int l, int r, int ele)
```
Comment (Monal5031) [inline binary_search/binary_search.c]: name of this function should be `binary_search_recursive` to show that it is a recursive solution and also to follow guidelines for cpp.

File: binary_search/binary_search.c [hunk 6]
```diff
@@ -5,6 +5,37 @@
...
+ {
+    if (r >= l)
+    {
+         int mid = l + (r - l)/2;    //calculate mid point of the array
+  
+         // If the element is present at the middle itself
+         
+         if (ar[mid] == ele){  return mid;   }
```
Comment (Monal5031) [inline binary_search/binary_search.c]: Please bring `return mid;` to next line. Also do the same in other places.

**Comment by HarshKhatore** [comment]
@Monal5031 Done with the correction. Please review.

**Comment by Monal5031** [comment]
@HarshKhatore Rather than @mentioning here, its better mention the reviewer after updates on [gitter channel](https://gitter.im/iiitv/algos) otherwise the pr update might not get noticed in many mails 😃

**Comment by HarshKhatore** [comment]
@Monal5031 Thanks for telling me this.

File: binary_search/binary_search.c [hunk 7]
```diff
@@ -5,6 +5,40 @@
...
+ 
+/* function to perform recursive binary search
+ * input parameters:
+ * @ar - user-input array.
+ * @l - left index of array.
+ * @r - right index of array.
+ * @ele - element to be searched.
+ */
+int recursive_binary_search(const int *ar, int l, int r, int ele)
+{
```
Comment (Monal5031) [inline binary_search/binary_search.c]: This opening bracket should start after the function name as `func_name(...) {` as per guidelines mentioned. Please change this for all conditional and looping statements too as `if (...) {` and `for (...) {`

File: binary_search/binary_search.c [hunk 8]
```diff
@@ -5,6 +5,40 @@
...
+	if (r >= l)
+    	{
+		int mid = l + (r - l)/2;    //calculate mid point of the array
+		
+		// If the element is present at the middle itself
+		if (ar[mid] == ele){
```
Comment (Monal5031) [inline binary_search/binary_search.c]: Please add a whitespace between `)` and `{` same for other places in code.

File: binary_search/binary_search.c [hunk 9]
```diff
@@ -25,15 +59,21 @@ int binary_search(const int *arr, int arr_size, int search_element) {
 int main() {
 	int arr[] = {1, 5, 35, 112, 258, 324},
 		search_arr[] = {1, 35, 112, 324, 67},
-		pos, i;
+		pos, i, pos_rec;
 
 	for (i = 0 ; i < 5 ; i ++) {
```
Comment (Monal5031) [inline binary_search/binary_search.c]: Please remove the space before increment. Its not needed.

File: binary_search/binary_search.c [hunk 10]
```diff
@@ -5,6 +5,39 @@
...
+		
+		// If the element is present at the middle itself
+		if (ar[mid] == ele) {  
+			return mid;   
+		}
+		
+		// If element is smaller than mid, then it can only be present
+		// in left subarray
+		if (ar[mid] > ele) { 
+			return binarySearch(ar, l, mid-1, ele);
```
Comment (Monal5031) [inline binary_search/binary_search.c]: Whitespace around binary operator, also at line 33

File: binary_search/binary_search.c [hunk 11]
```diff
@@ -5,6 +5,39 @@
...
  */
+ 
+/* function to perform recursive binary search
+ * input parameters:
+ * @ar - user-input array.
+ * @l - left index of array.
+ * @r - right index of array.
+ * @ele - element to be searched.
+ */
+int recursive_binary_search(const int *ar, int l, int r, int ele) {
```
Comment (singhpratyush) [inline binary_search/binary_search.c]: Please rename this as `binary_search_recursive`.

Comment (singhpratyush) [inline binary_search/binary_search.c]: Also, every variable should have their own declaration statement. Please split these.

File: binary_search/binary_search.c [hunk 12]
```diff
@@ -25,15 +58,21 @@ int binary_search(const int *arr, int arr_size, int search_element) {
...
-		pos, i;
+		pos, i, pos_rec;
 
-	for (i = 0 ; i < 5 ; i ++) {
+	for (i = 0 ; i < 5 ; i++) {
 		pos = binary_search(arr, 6, search_arr[i]);
-
+		pos_rec = recursive_binary_search(arr, 0, 5, search_arr[i]);
+		
 		if (pos >= 0)
```
Comment (singhpratyush) [inline binary_search/binary_search.c]: Wrap all the conditional statements' body in a block.

File: binary_search/binary_search.c (hunk 11 above)
Comment (singhpratyush) [inline binary_search/binary_search.c]: I was referring to line 59 here, don't know why it's showing up here.

File: binary_search/binary_search.c [hunk 13]
```diff
@@ -23,17 +56,29 @@ int binary_search(const int *arr, int arr_size, int search_element) {
...
 			printf("%d found at index %d.\n", search_arr[i], pos);
-		else
+		}
+		else {
+			printf("%d not found.\n", search_arr[i]);
+		}
+		
+		pos = binary_search_recursive(arr, 0, 5, search_arr[i]);
+		
+		if(pos_rec >= 0) {
```
Comment (singhpratyush) [inline binary_search/binary_search.c]: Where is this variable defined?

Please introduce a space after `if`.

File: binary_search/binary_search.c [hunk 14]
```diff
@@ -23,17 +56,29 @@ int binary_search(const int *arr, int arr_size, int search_element) {
...
+	int search_arr[] = {1, 35, 112, 324, 67},
+	int pos;
+	int i;
 
-	for (i = 0 ; i < 5 ; i ++) {
+	for (i = 0 ; i < 5 ; i++) {
 		pos = binary_search(arr, 6, search_arr[i]);
-
-		if (pos >= 0)
+
```
Comment (singhpratyush) [inline binary_search/binary_search.c]: Trailing whitespace.

File: binary_search/binary_search.c [hunk 15]
```diff
@@ -23,17 +56,29 @@ int binary_search(const int *arr, int arr_size, int search_element) {
 }
 
 int main() {
-	int arr[] = {1, 5, 35, 112, 258, 324},
-		search_arr[] = {1, 35, 112, 324, 67},
-		pos, i;
+	int arr[] = {1, 5, 35, 112, 258, 324};
+	int search_arr[] = {1, 35, 112, 324, 67},
+	int pos;
```
Comment (singhpratyush) [inline binary_search/binary_search.c]: Resolve this issue.

File: binary_search/binary_search.c [hunk 16]
```diff
@@ -23,17 +56,28 @@ int binary_search(const int *arr, int arr_size, int search_element) {
...
+		
+		int pos;
 		pos = binary_search(arr, 6, search_arr[i]);
-
-		if (pos >= 0)
+		if (pos >= 0) {
 			printf("%d found at index %d.\n", search_arr[i], pos);
-		else
+		}
+		else {
```
Comment (singhpratyush) [inline binary_search/binary_search.c]: `else` would go in previous line - 
```c
if (...) {
    ...
} else {
    ...
}
```

**Comment by HarshKhatore** [comment]
@singhpratyush Done the required changes.

**Comment by HarshKhatore** [comment]
@singhpratyush Dona all the required changes.

**Pull request merged**